src/osp/id_registry.h

```diff
@@ -0,0 +1,341 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "hierarchical_bitset.h"
+
+#include <cassert>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace osp
+{
+
+template<class TYPE_T>
+constexpr TYPE_T id_null() noexcept
+{
+    using underlying_t = typename std::underlying_type_t<TYPE_T>;
+    return TYPE_T(std::numeric_limits<underlying_t>::max());
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief Generates reusable sequential IDs
+ */
+template<typename ID_T, bool NO_AUTO_RESIZE = false>
+class IdRegistry
+{
+    using id_int_t = std::underlying_type_t<ID_T>;
+
+public:
+
+    IdRegistry() = default;
+    IdRegistry(size_t capacity) { reserve(capacity); };
+
+    /**
+     * @brief Create a single ID
+     *
+     * @return A newly created ID
+     */
+    [[nodiscard]] ID_T create()
+    {
+        ID_T output;
+        create(&output, 1);
+        return output;
+    }
+
+    /**
+     * @brief Create multiple IDs
+     *
+     * @param out   [out] Iterator out
+     * @param count [in] Number of IDs to generate
+     */
+    template<typename IT_T>
+    void create(IT_T out, size_t count);
```

Comment:
I'm not sure this is the ideal interface.

Generally, you wouldn't mix iterators and counts like this.

Instead, I recommend making two (maybe three?) versions.

1) Pointer + size_t -- requires contiguous.
2) Begin + end iterators -- does not require contiguous.
3) For the case where you create only a single item, make the create() function return that item (Which ou've done on line 65).

Reply:
I was kind of inspired by std::copy's output iterator. I'm aware that this is a bit weird, but it's tailored around how the ID registry will be used in code soon. If first and last iterators are used, then most of the time, last would just be "first + count".